spec/spec_helper_acceptance.rb

```diff
@@ -1,30 +1,128 @@
 require 'beaker-rspec/spec_helper'
 require 'beaker-rspec/helpers/serverspec'
 
+def stop_firewall_on(host)
+  case host['platform']
+  when /debian/
+    on host, 'iptables -F'
+  when /fedora|el-7/
+    on host, puppet('resource', 'service', 'firewalld', 'ensure=stopped')
+  when /el|centos/
+    on host, puppet('resource', 'service', 'iptables', 'ensure=stopped')
+  when /ubuntu/
+    on host, puppet('resource', 'service', 'ufw', 'ensure=stopped')
+  else
+    logger.notify("Not sure how to clear firewall on #{host['platform']}")
+  end
+end
+
+# Project root
+PROJ_ROOT = File.expand_path(File.join(File.dirname(__FILE__), '..'))
+
 unless ENV['BEAKER_provision'] == 'no'
-  hosts.each do |host|
+  # Work-around for BKR-262
+  @logger = logger
+
+  # Setup repositories if using a specific SHA
+  if ENV['SHA']
+    hosts.each do |host|
+      install_puppetlabs_release_repo(host)
+      install_puppetlabs_dev_repo(host, 'puppet', ENV['SHA'])
+    end
+
+    if master
+      install_package master, 'puppet-server'
+      master['use-service'] = true
+    end
+  end
+end
+
+def parser_opts
+  # Configuration only needed on 3.x master
+  {:main => {:stringify_facts => false, :parser => 'future'}}
+end
+
+def setup_puppet(agent_run = false)
+
+  agents.each do |agent|
     # Install Puppet
-    if host.is_pe?
+    if ENV['SHA']
+      install_package agent, 'puppet'
+    elsif agent.is_pe?
       install_pe
     else
       install_puppet
     end
+
+    configure_puppet_on(agent, parser_opts)
   end
-end
 
-RSpec.configure do |c|
-  # Project root
-  proj_root = File.expand_path(File.join(File.dirname(__FILE__), '..'))
+  if master and agent_run
+    # Initialize SSL
+    hostname = on(master, 'facter hostname').stdout.strip
+    fqdn = on(master, 'facter fqdn').stdout.strip
 
-  # Readable test descriptions
-  c.formatter = :documentation
+    if master.use_service_scripts?
+      step "Ensure puppet is stopped"
+      # Passenger, in particular, must be shutdown for the cert setup steps to work,
+      # but any running puppet master will interfere with webrick starting up and
+      # potentially ignore the puppet.conf changes.
+      on(master, puppet('resource', 'service', master['puppetservice'], "ensure=stopped"))
+    end
 
-  # Configure all nodes in nodeset
-  c.before :suite do
-    # Install module and dependencies
-    puppet_module_install(:source => proj_root, :module_name => 'agent_upgrade')
+    step "Clear SSL on all hosts"
     hosts.each do |host|
-      on host, puppet('module', 'install', 'puppetlabs-stdlib'), { :acceptable_exit_codes => [0,1] }
+      stop_firewall_on host
+      ssldir = on(host, puppet('agent --configprint ssldir')).stdout.chomp
+      on(host, "rm -rf '#{ssldir}'")
+    end
+
+    step "Master: Start Puppet Master" do
```

Comment:
Some of this master/agent co-ordination would be awesome in puppet_helpers (whether that's in Beaker or split into a Beaker helper module).